### PR TITLE
Improve ES output error insights

### PR DESCRIPTION
**Pull request by ChrsMark**
## What does this PR do?
This PR improves ES output insights by:
- do not record EOF as error  in `StatsDialer`
- wrap `StatsDialer`s  with a new dialer wrapper that logs errors

## Why is it important?
To provide [better insights](https://github.com/elastic/sdh-beats/issues/937) on Output errors.

### CODE REVIEW

File: libbeat/common/transport/logging.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+		return &loggingConn{c, logger}
+	})
+}
+
+func (l *loggingConn) Read(b []byte) (int, error) {
+	n, err := l.Conn.Read(b)
+	if err != nil {
+		// ignore EOF error
+		if err == io.EOF {
+			err = nil
```
Comment (urso) [inline libbeat/common/transport/logging.go:43]: Do not eat EOF.

File: libbeat/common/transport/logging.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+	}
+	return n, err
+}
+
+func (l *loggingConn) Write(b []byte) (int, error) {
+	n, err := l.Conn.Write(b)
+	if err != nil {
+		// ignore EOF error
+		if err == io.EOF {
+			err = nil
```
Comment (urso) [inline libbeat/common/transport/logging.go:56]: do not eat EOF. We want to report only.

File: libbeat/common/transport/stats.go [hunk 3]
```diff
@@ -43,7 +44,12 @@ func StatsDialer(d Dialer, s IOStatser) Dialer {
 func (s *statsConn) Read(b []byte) (int, error) {
 	n, err := s.Conn.Read(b)
 	if err != nil {
-		s.stats.ReadError(err)
+		// ignore EOF error
+		if err == io.EOF {
+			err = nil
+		} else {
+			s.stats.ReadError(err)
+		}
```
Comment (urso) [inline libbeat/common/transport/stats.go:52]: how about: `if err != nil && err != io.EOF && err != context.Canceled { s.stats.ReadError(err) }` ?

File: libbeat/esleg/eslegclient/connection.go [hunk 4]
```diff
@@ -110,6 +110,9 @@ func NewConnection(s ConnectionSettings) (*Connection, error) {
 	if st := s.Observer; st != nil {
 		dialer = transport.StatsDialer(dialer, st)
 		tlsDialer = transport.StatsDialer(tlsDialer, st)
+		logger := logp.NewLogger("esclientleg")
+		dialer = transport.LoggingDialer(dialer, logger)
+		tlsDialer = transport.LoggingDialer(tlsDialer, logger)
```
Comment (urso) [inline libbeat/esleg/eslegclient/connection.go:115]: We always want the logging. Lets move the `LoggingDialer` out of the if statement.

**Review by urso — CHANGES_REQUESTED** [review]
Is there a reason for not instrumenting `statsConn.Write`?

**Comment by ChrsMark** [comment]
> Is there a reason for not instrumenting `statsConn.Write`?

Tuned `Write` too. Let me know you think.

File: libbeat/common/transport/logging.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+	"github.com/elastic/beats/v7/libbeat/logp"
+)
+
+type loggingConn struct {
+	net.Conn
+	logger *logp.Logger
+}
+
+func LoggingDialer(d Dialer, logger *logp.Logger) Dialer {
+	return ConnWrapper(d, func(c net.Conn) net.Conn {
```
Comment (urso) [inline libbeat/common/transport/logging.go:33]: One more ask: let's make use of structured logging, so we can tell which host the log output is to be associated with. e.g.

```
return DialFunc(func(network, address string) (net.Conn, error) {
  logger := logger.With("network", network, "address", address)
  conn, err := d.Dial(network, address)
  ...
})
```

The address includes `<hostname|ip>:<port>`.

File: libbeat/common/transport/logging.go [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+	net.Conn
+	logger *logp.Logger
+}
+
+func LoggingDialer(d Dialer, logger *logp.Logger) Dialer {
+	return DialerFunc(func(network, addr string) (net.Conn, error) {
+		logger := logger.With("network", network, "address", addr)
+		c, err := d.Dial(network, addr)
+		if err != nil {
+			logger.Errorf("error dialing", err)
```
Comment (urso) [inline libbeat/common/transport/logging.go:37]: Missing `%v` or do not use `Errorf`.

nit; Log messages should be a sentence and start with uppercase.

File: libbeat/common/transport/logging.go [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+		logger.Debugf("Completed dialing successfully")
+		return &loggingConn{c, logger}, nil
+	})
+}
+
+func (l *loggingConn) Read(b []byte) (int, error) {
+	n, err := l.Conn.Read(b)
+	if err != nil && err != io.EOF {
+		l.logger.Debugf("error reading from connection: %v", err)
```
Comment (urso) [inline libbeat/common/transport/logging.go:49]: Uppercase

File: libbeat/common/transport/logging.go [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+	}
+	return n, err
+}
+
+func (l *loggingConn) Write(b []byte) (int, error) {
+	n, err := l.Conn.Write(b)
+	if err != nil && err != io.EOF {
+		l.logger.Debugf("error writing to connection: %v", err)
```
Comment (urso) [inline libbeat/common/transport/logging.go:57]: Uppercase

**Review by urso — APPROVED** [review]

**Comment by ChrsMark** [comment]
Failing tests seem to be unrelated (hitting issue reported at https://github.com/elastic/beats/issues/25956). Merging this one.

**Pull request merged**